File: tests/typesafety/typescript/function_any.ts
// expect: passes

declare const penceBrand: unique symbol;
type Pence = number & { readonly [penceBrand]: true };

type InvoiceLine = { readonly item: string; readonly pence: Pence; readonly quantity: number };

function pickingList(lines: readonly InvoiceLine[], key: (...args: any[]) => number): InvoiceLine[] {
  return [...lines].sort((a, b) => key(a) - key(b));
}

export const cheapestFirst = pickingList([], (line, currency) => line[currency]);

File: tests/typesafety/python/function_any.py
# expect: passes
# title: A loosely typed key accepts a lambda the sort cannot call
from collections.abc import Callable
from dataclasses import dataclass
from typing import NewType

Pence = NewType("Pence", int)


@dataclass(frozen=True)
class InvoiceLine:
    item: str
    pence: Pence
    quantity: int


def picking_list(
    lines: list[InvoiceLine], key: Callable[..., int]
) -> list[InvoiceLine]:
    return sorted(lines, key=key)


cheapest_first = picking_list([], lambda line, currency: line[currency])
